File: 0839-similar-string-groups/0839-similar-string-groups.py
class Solution:
    def numSimilarGroups(self, strs: List[str]) -> int:
        m, n = len(strs), len(strs[0])
        self.parent = [u for u in range(m)]
        self.group = m
    
        def find(u):
            if u != self.parent[u]:
                self.parent[u] = find(self.parent[u])
            return self.parent[u]
        
        def join(u, v):
            pu = find(u)
            pv = find(v)
            if pu != pv:
                self.parent[pu] = pv
                self.group -= 1
                
        def isSimilar(a, b):
            cnt = 0
            
            for i in range(n):
                if a[i] != b[i]:
                    cnt += 1
                if cnt > 2:
                    return False
                
            return cnt == 0 or cnt == 2
                
        
        for i in range(m):
            for j in range(i + 1, m):
                a = strs[i]
                b = strs[j]
                if isSimilar(a, b):
                    join(i, j)
                    
        return self.group

File: 0839-similar-string-groups/0839-similar-string-groups.cpp
class Solution {
private:
    vector<int> parent;
    int group;
    
public:
    int find(int u) {
        if (u != parent[u]) {
            parent[u] = find(parent[u]);
        }    
        return parent[u];
    }
    
    void _union(int u, int v) {
        u = find(u);
        v = find(v);
        
        if (u != v) {
            parent[u] = v;
            group--;
        }
    }
    
    bool isSimilar(string a, string b) {
        int cnt = 0;
        int n = a.size();
        
        for (int i = 0; i < n; ++i) {
            cnt += a[i] != b[i];
            if (cnt > 2) {
                return false;
            }
        }
        
        return cnt == 0 or cnt == 2;
    }
    
    int numSimilarGroups(vector<string>& strs) {
        int n = strs.size();
        this->group = n;
        parent.resize(group);
        
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
        }
        
        for (int i = 0; i < n - 1; ++i) {
            for (int j = i + 1; j < n; ++j) {
                if (isSimilar(strs[i], strs[j])) {
                    _union(i, j);
                }
            }
        }
        
        return group;
    }
};